components/tabs/src/sync/record.rs

```diff
@@ -27,7 +28,163 @@ pub struct TabsRecord {
 impl TabsRecord {
     #[inline]
     pub fn from_payload(payload: sync15::Payload) -> Result<Self> {
-        let record: TabsRecord = payload.into_record()?;
-        Ok(record)
+        let client_name: String = payload.data["clientName"]
+            .as_str()
+            .unwrap_or_default()
+            .to_string();
+        let tabs: Vec<TabsRecordTab> = payload.data["tabs"]
+            .as_array()
+            .unwrap_or(&Vec::new())
+            .iter()
+            .map(|x| {
+                let tabs_obj: Map<String, JsonValue> = x.as_object().unwrap_or(&Map::new()).clone();
+                let title: String = tabs_obj["title"].as_str().unwrap_or_default().to_string();
+                let url_history: Vec<String> = tabs_obj["urlHistory"]
+                    .as_array()
+                    .unwrap_or(&Vec::new())
+                    .iter()
+                    .map(|u| u.as_str().unwrap_or_default().to_string())
+                    .collect::<Vec<_>>();
+                let icon: Option<String> = tabs_obj
+                    .get("icon")
+                    .map(|i| i.as_str().unwrap_or_default().to_string());
+                let last_used = parse_last_used(&tabs_obj["lastUsed"]);
+
+                TabsRecordTab {
+                    title,
+                    url_history,
+                    icon,
+                    last_used,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let ttl: u32 = payload.data["ttl"].as_u64().unwrap_or_default() as u32;
+
+        Ok(TabsRecord {
+            id: payload.id.to_string(),
+            client_name,
+            tabs,
+            ttl,
+        })
+    }
+}
+
+fn parse_last_used(last_used_val: &JsonValue) -> u64 {
+    // In order to support older clients, we are handling `last used` values that
+    // are either floats, integers, or stringified floats or integers. We attempt to
+    // represent `last used` as a float before converting it to an integer. If that
+    // operation isn't successful, we try converting `last used` to an integer directly.
+    // If that isn't successful, the returned value will be zero.
+
+    if last_used_val.is_string() {
+        let l = last_used_val.as_str().unwrap_or_default();
+
+        match l.parse::<f64>() {
+            Ok(f) => f.trunc() as u64,
+            Err(_) => l.parse::<u64>().unwrap_or_default(),
+        }
+    } else {
+        match last_used_val.as_f64() {
+            Some(f) => f.trunc() as u64,
+            None => last_used_val.as_u64().unwrap_or_default(),
```

Comment:
As above, this will now ignore invalid `lastUsed` fields and replace with the default value of zero.